qa/full-cluster-restart/src/test/java/org/elasticsearch/upgrades/FullClusterRestartIT.java

```diff
@@ -1248,4 +1255,37 @@ public void testPeerRecoveryRetentionLeases() throws IOException {
             RetentionLeaseUtils.assertAllCopiesHavePeerRecoveryRetentionLeases(client(), index);
         }
     }
+
+    /**
+     * Tests that with or without soft-deletes, we should perform an operation-based recovery if there were some but not too many
+     * uncommitted documents before we restart the cluster. This is important when we move from translog based to retention leases
```

Comment:
how do you quantify this "not too many"? It assumes no background merging that creates a new safe commit? I wonder how this is ensured by this test

Reply:
Good question. I thought a bit more about this issue. I believe we turn off the translog retention too early with https://github.com/elastic/elasticsearch/pull/45473. During a rolling upgrade, a replica can be allocated to a 7.4+ node, while the primary is still on a pre-7.4 node. In this case, the replica disables the translog retention, although no PRRL is installed. We then are likely to perform a file-based recovery when upgrading the node with the primary copy.

I think we should only disable the translog retention after all shard copies started and have PPRL in place (or the index_created_version >= 7.4). Moreover, if a recovering replica does not have a PPRL, we should fallback using translog. This should also offer an upgrade path for indices without soft-deletes. WDYT?

Reply:
ok, let's try this out